### ISSUE TITLE
[all] Add macOS support

### ISSUE THREAD

**Issue by cbenhagen** [FirebaseExtended/flutterfire#1653]
From the Firebase iOS SDK readme:
> Thanks to contributions from the community, FirebaseABTesting, FirebaseAuth, FirebaseCore, FirebaseDatabase, FirebaseMessaging, FirebaseFirestore, FirebaseFunctions, FirebaseRemoteConfig, and FirebaseStorage now compile, run unit tests, and work on macOS and tvOS.

This should make it quite easy to support most Firebase plugins on macOS. If noone is already working on this I will try to experiment with `firebase_core`, `firebase_auth`, `cloud_functions` and `cloud_firestore`. Help and guidance very much appreciated.

How do you feel about converting those plugins to Swift at the same time? It's not really needed but I would feel much more comfortable with that language.

/cc @ditman @hterkelsen @collinjackson

**Comment by cbenhagen** [comment]
PRs for `firebase_core`, `firebase_auth`, `cloud_functions` and `cloud_firestore` are created. They are all tested in our app but lack unit tests. I will try to add them in the coming days.

They are 99% identical to their iOS siblings. I currently added them as `macos` folders next to the `ios` implementation. Federation would be an option but I think it would be ideal if we could find a way to even keep both implementations in the same file to prevent them from drifting apart. What would be the best approach for that?

**Comment by ditman** [comment]
> Federation would be an option but I think it would be ideal if we could find a way to even keep both implementations in the same file to prevent them from drifting apart. What would be the best approach for that?

@cbenhagen What about extracting the common ios and macos parts to a shared package that gets dependend on by both the ios plugin (in the core package) and the macos version (in its own, federated, package)?

**Comment by laxxxguy** [comment]
how is it going? and how do we add osx support for firebase, is it just like ios? i mean info.plist file.

**Comment by abjorck** [comment]
@cbenhagen Great work, happy to see progress on this! 
Any plans to do `firebase_database` as well? :) (I want to access some "legacy" data there.) 
I gave it a bit of a shot before, but I'm not so good with xcode, it seems... 
Maybe I can learn enough from your PRs otherwise

**Comment by cbenhagen** [comment]
@abjorck I can't promise anything but I will try and see how far I get. Will add `firebase_crashlytics` and `firebase_messaging` for sure as I am also using them.

**Comment by viktor-morin** [comment]
> @abjorck I can't promise anything but I will try and see how far I get. Will add `firebase_crashlytics` and `firebase_messaging` for sure as I am also using them.

Is the crashlytics plugging working for iOS? Me and some other can't get it to work following the implementation guid. See issue described:

[StackOverflow: flutter-flutter-firebase-crashlytics-not-showing-for-ios](https://stackoverflow.com/questions/60111182/flutter-flutter-firebase-crashlytics-not-showing-for-ios)

**Comment by cbenhagen** [comment]
@viktor-morin please stop hijacking other issues.

**Comment by viktor-morin** [comment]
> @viktor-morin please stop hijacking other issues.

I'm not trying yo hiijacking anything, I'm trying to solve an issue. So this isn't related either? So this is just for macOS and not iOS? I was hoping solving macOS would solve iOS as well. My bad.

**Comment by cbenhagen** [comment]
Fixing your issue for iOS would most likely also fix it for macOS as the code is shared. This issue here was created to track overall progress of macOS implementations of all firebase plugins and to discuss topics which touch them all like sharing code between iOS and macOS.

Please keep the communication about your specific issue in the issue itself unless you have good reasons to believe that another issue is related. If it just mentions iOS as well it does not mean that the issue is in any way connected.

**Comment by viktor-morin** [comment]
Manage to solve it for MAC, apparently you need in the Firebase Console, for iOS press "Enable SDK"

**Step 1: Set up Crashlytics in the Firebase console**
Click Crashlytics in the left-hand nav panel of the Firebase console.

If your Firebase project has multiple apps registered in it, select the app you just added from the dropdown next to Crashlytics in the top bar of the console.

**Click Add SDK.**

[https://firebase.google.com/docs/crashlytics/get-started](https://firebase.google.com/docs/crashlytics/get-started)

**Comment by EliasDeuss** [comment]
@cbenhagen do you have any plans to add firebase_analytics support?

**Comment by cbenhagen** [comment]
This is blocked by https://github.com/firebase/firebase-ios-sdk/issues/1732

**Comment by Salakar** [comment]
Hey all, just an FYI `firebase_database` MacOS support is landing shortly via #3021

**Comment by Salakar** [comment]
Hey all, another FYI 🙃  `firebase_remote_config` MacOS support is landing shortly via #3066
